File: Day1/go/solve.go
package main

import (
	"fmt"

	"io/ioutil"
	"log"
	"math"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

func getData() ([]int, []int) {
	var listA []int
	var listB []int
	fileContent, err := ioutil.ReadFile("../input.txt")
	if err != nil {
		log.Fatal(err)
	}
	lines := strings.Split(string(fileContent), "\n")
	re := regexp.MustCompile("[0-9]+")
	for _, line := range lines {
		abc := re.FindAllString(line, -1)

		A, err := strconv.Atoi(abc[0])
		if err != nil {
			panic(err)
		}
		listA = append(listA, A)

		B, err := strconv.Atoi(abc[1])
		if err != nil {
			panic(err)
		}
		listB = append(listB, B)

	}
	return listA, listB

}

func stringToInt(arr []string) []int {
	var ret = make([]int, len(arr))
	for idx, i := range arr {
		j, err := strconv.Atoi(i)
		if err != nil {
			panic(err)
		}
		ret[idx] = j
	}
	return ret
}

func part1(list1 []int, list2 []int) int {
	distance := 0
	for idx, i := range list1 {
		distance += int(math.Abs(float64(i) - float64(list2[idx])))
	}
	return distance

}
func countValue(arr []int) map[int]int {
	//Create a   dictionary of values for each element
	dict := make(map[int]int)
	for _, num := range arr {
		dict[num] = dict[num] + 1
	}
	return dict
}

func part2(A []int, B []int) int {
	counts := countValue(B)

	simillarityScore := 0
	for _, vals := range A {
		if val, ok := counts[vals]; ok {

			simillarityScore += (vals * val)
		}
	}
	return simillarityScore
}

func main() {
	var listA, listB = getData()
	sort.Slice(listA, func(i, j int) bool {
		return listA[j] > listA[i]
	})
	sort.Slice(listB, func(i, j int) bool {
		return listB[j] > listB[i]
	})

	fmt.Println(part1(listA, listB)) // 1590491
	fmt.Println(part2(listA, listB)) // 22588371

}

File: Day1/php/solve.php
<?php
$file  = fopen("../input.txt", "r");
$lines = [];
$listA = [];
$listB = [];
while (! feof($file)) {
    $lines[] = fgets($file);
}
fclose($file);

foreach ($lines as $line) {
    preg_match_all('/(\d+)/', $line, $matches);
    $listA[] = $matches[0][0];
    $listB[] = $matches[0][1];
}

function part_1($list1, $list2)
{
    sort($list1);
    sort($list2);
    $lens     = count($list1);
    $indx     = 0;
    $distance = 0;
    while ($indx < $lens) {
        $distance += abs($list1[$indx] - $list2[$indx]);

        $indx++;
    }
    return $distance;
}

function part_2($list1, $list2)
{
    $lens              = count($list1);
    $indx              = 0;
    $simillarity_score = 0;

    while ($indx < $lens) {
        $curr = $list1[$indx];
        $simillarity_score += array_key_exists($curr, $list2) ? $curr * $list2[$curr] : 0;
        $indx++;
    }
    return $simillarity_score;

}

print(part_1($listA, $listB)); //1590491
$countValues = array_count_values($listB);
print("\n");
print(part_2($listA, $countValues)); //22588371
